src/regex/pcre2.cr

```diff
@@ -0,0 +1,183 @@
+      res
+    else
+      message = String.new(256) do |buffer|
+        bytesize = LibPCRE2.get_error_message(errorcode, buffer, 256)
+        {bytesize, 0}
+      end
+      yield "#{message} at #{erroroffset}"
+    end
+  end
+
+  private def pcre2_options(options)
+    flag = 0
+    options.each do |option|
+      flag |= case option
+              when .ignore_case?   then LibPCRE2::CASELESS
+              when .multiline?     then LibPCRE2::DOTALL | LibPCRE2::MULTILINE
+              when .extended?      then LibPCRE2::EXTENDED
+              when .anchored?      then LibPCRE2::ANCHORED
+              when .utf_8?         then LibPCRE2::UTF
+              when .no_utf8_check? then LibPCRE2::NO_UTF_CHECK
+              when .dupnames?      then LibPCRE2::DUPNAMES
+              when .ucp?           then LibPCRE2::UCP
+              else
+                raise "unreachable"
+              end
+    end
+    flag
+  end
+
+  def finalize
+    {% unless flag?(:interpreted) %}
+      LibPCRE2.code_free @re
+    {% end %}
+  end
+
+  protected def self.error_impl(source)
+    code = PCRE2.compile(source, LibPCRE2::UTF | LibPCRE2::NO_UTF_CHECK | LibPCRE2::DUPNAMES | LibPCRE2::UCP) do |error_message|
+      return error_message
+    end
+
+    LibPCRE2.code_free code
+
+    nil
+  end
+
+  private def pattern_info(what)
+    value = uninitialized UInt32
+    pattern_info(what, pointerof(value))
+    value
+  end
+
+  private def pattern_info(what, where)
+    ret = LibPCRE2.pattern_info(@re, what, where)
+    if ret != 0
+      raise "error pattern_info #{what}: #{ret}"
+    end
+  end
+
+  private def name_table_impl
+    lookup = Hash(Int32, String).new
+
+    each_capture_group do |capture_number, name_entry|
+      lookup[capture_number] = String.new(name_entry.to_unsafe + 2)
+    end
+
+    lookup
+  end
+
+  # :nodoc:
+  def each_capture_group
+    name_table = uninitialized UInt8*
+    pattern_info(LibPCRE2::INFO_NAMETABLE, pointerof(name_table))
+
+    name_entry_size = pattern_info(LibPCRE2::INFO_NAMEENTRYSIZE)
+
+    name_count = pattern_info(LibPCRE2::INFO_NAMECOUNT)
+    name_count.times do
+      capture_number = (name_table[0] << 8) | name_table[1]
+
+      yield capture_number, Slice.new(name_table, name_entry_size)
+
+      name_table += name_entry_size
+    end
+  end
+
+  private def capture_count_impl
+    pattern_info(LibPCRE2::INFO_CAPTURECOUNT).to_i32
+  end
+
+  private def match_impl(str, byte_index, options)
+    match_data = match_data(str, byte_index, options) || return
+
+    ovector = LibPCRE2.get_ovector_pointer(match_data)
+    ovector_count = LibPCRE2.get_ovector_count(match_data)
+
+    ::Regex::MatchData.new(self, @re, str, byte_index, ovector, ovector_count.to_i32 - 1)
+  end
+
+  private def matches_impl(str, byte_index, options)
+    if match_data = match_data(str, byte_index, options)
+      true
+    else
+      false
+    end
+  end
+
+  class_getter general_context do
```

Comment:
shouldn't this be private (or protected)?

Reply:
The entire `Regex::PCRE2` namespace is `:nodoc:` so I think visibility doesn't matter much. I suppose it could be protected, though.

Reply:
yes, it's just nitpicking, I was wondering if there were a reason to leave it public